### PR TITLE
[C++20] [Modules] Introduce thin BMI

### LINKED ISSUE #71034: [C++20] [Modules] Try to introduce thin BMI to exclude the things not necessary in an interface
Let's call the current BMI for named modules as fat BMI. Since it contains all the information needed to compile the BMI itself to an object file. And this is the foundation of 2 phase compilations.

However, it is a waste of time and space if we're using one phase compilations. 

Then let's call the BMI which doesn't contain all the information to produce an object file, but is able to satisfy the requirement of consumers of the modules as thin BMI. Then it is clear that we should use thin BMI then fat BMI in one phase compilation.

Also in 2 phase compilations, the thin BMI may be helpful too if the build systems want to use a compare-and-swap strategy to generate the BMI based on the BMIDeclsHash (See https://github.com/llvm/llvm-project/issues/71618).

### PR DESCRIPTION
**Pull request by ChuanqiXu9**
Close https://github.com/llvm/llvm-project/issues/71034

This patch introduces thin BMI, which doesn't contain the definitions of functions and variables if its definitions won't contribute to the ABI.

Testing is a big part of the patch. We want to make sure the thin BMI contains the same behavior with the existing and relatively stable fatBMI. This is pretty helpful for further reduction.

For user interfeaces, this patch introduces `-fthinBMI-output=` arguments to specify the position of thin BMI. This should be used when compiling a single module unit.

The design is helpful to use thin BMI in two phase compilations too. With thin BMI, In two phase compilations, we'll generate 2 BMIs, one thin BMI for being used by consumers, one fat BMI for compiling itself to object files. Maybe it sounds confusing to have 2 BMIs for one module unit. But only the thin BMI will be the BMI we're talking about generally and the fat BMI is only visible by the module unit itself.

With one phase compilation, we may find the behavior of `-fthinBMI-output=` is pretty similar with `-fmodule-output=`, except one generating thin BMI and the other generating fat BMI. The design here is based on 2 things:
(1) The serialization of C++ is pretty complex. We can't be sure we're handling every detail correctly in the every beginning.
(2) The fat BMI is relatively widely used and relatively stable. So it looks not good to replace the fat BMI immediately with thin BMI.

But, of course, in the end of the day, we want the consumers to use the thin BMI only. When that day comes, the `-fmodule-output=` will be an alias to `-fthinBMI-output=`.

Another design choice is to reuse `-fmodule-output=` and introduce a flag `-femit-thin-BMI`. Then `-femit-thin-BMI -fmodule-output=` will have the same effect with `-fthinBMI-output=` now.
The flag `-femit-thin-BMI` should be opt-in now and opt-off later and finally deprecated.

The roadmap for thin BMI in my mind is:

(1) In clang18, release thin BMI and mark it as experimental. Also encourage users and build systems to try this new mode. (2) In clang19 or clang20 (based on the issue feedbacks), remove the experimental mark for thin BMI and mark fat BMI as deprecated to be used by consumers.
(3) In clang21 or clang22, error out if we found the users are trying to import a fat BMI.

CC: @mathstuf

### CODE REVIEW

**Comment by mathstuf** [comment]
It was suggested to avoid "thin" and "fat" terminology (I started using it with LTO as a precedent for the terms), but I suppose something more descriptive could be selected.

**Comment by mathstuf** [comment]
I'll also note that my prior usage was in reference to embedding transitive BMI references in the BMI, not about including function definitions.

**Comment by ChuanqiXu9** [comment]
> It was suggested to avoid "thin" and "fat" terminology (I started using it with LTO as a precedent for the terms), but I suppose something more descriptive could be selected.

Any suggestions? My english is not good, you know.

> I'll also note that my prior usage was in reference to embedding transitive BMI references in the BMI, not about including function definitions.

May you elaborate on this? I don't understand since I feel they are unrelated. Are you saying: https://github.com/llvm/llvm-project/issues/62707? I feel they are orthogonal.

File: clang/lib/Serialization/ASTWriterDecl.cpp [hunk 1]
```diff
@@ -270,6 +274,35 @@ namespace clang {
   };
 }
 
+bool clang::MayDefAffectABI(const Decl *D) {
+  if (auto *FD = dyn_cast<FunctionDecl>(D)) {
```
Comment (tbaederr) [inline clang/lib/Serialization/ASTWriterDecl.cpp:278]:
```suggestion
  if (const auto *FD = dyn_cast<FunctionDecl>(D)) {
```

File: clang/lib/Serialization/ASTWriterDecl.cpp [hunk 2]
```diff
@@ -270,6 +274,35 @@ namespace clang {
...
+      return true;
+
+    if (FD->isDependentContext())
+      return true;
+
+    if (FD->getTemplateSpecializationKind() == TSK_ImplicitInstantiation)
+      return true;
+  }
+
+  if (auto *VD = dyn_cast<VarDecl>(D)) {
```
Comment (tbaederr) [inline clang/lib/Serialization/ASTWriterDecl.cpp:294]:
```suggestion
  if (const auto *VD = dyn_cast<VarDecl>(D)) {
```

**Comment by iains** [comment]
perhaps in the context we can use "sparse" instead of "thin" - but I do not have a good short word replace 'fat'.

**Comment by tschuett** [comment]
It could also be a `surface BMI` as it only describes the surface of the module, but it does not contain its content.

**Comment by ChuanqiXu9** [comment]
The term `sparse` makes me wondering its format is not efficiency.. And for the term `surface`, ..., it may beyond my knowledge,  I feel it is odd but I can't give a concrete reason.

**Comment by ChuanqiXu9** [comment]
Let's discuss the higher level problems in https://discourse.llvm.org/t/rfc-c-20-modules-introduce-thin-bmi-and-decls-hash/74755

**Comment by dwblaikie** [comment]
The words probably don't need to be short.

Interface BMI
Implementation BMI

Seem like fine, accurate terms. I guess we could say "BMInterface" and "BMImplementation" but probably best to jus tgloss over "I" being "interface" and have "interface BMI" and "implementation BMI"

**Comment by ChuanqiXu9** [comment]
Since there is no meaningful review opinions here, let's continue it in a new and clean PR: https://github.com/llvm/llvm-project/pull/75894

**Comment by mathstuf** [comment]
> Since there is no meaningful review opinions here, let's continue it in a new and clean PR: https://github.com/llvm/llvm-project/pull/75894

Note that this leaves behind all who have hit "Subscribe" on this PR to keep tabs on things (I wish Github provided a way to mark-as-duplicate and merge the subscriber list). All of the references to this PR are now confused as well. What benefit does a "clean" PR provide?

**Comment by ChuanqiXu9** [comment]
> > Since there is no meaningful review opinions here, let's continue it in a new and clean PR: #75894
> 
> Note that this leaves behind all who have hit "Subscribe" on this PR to keep tabs on things (I wish Github provided a way to mark-as-duplicate and merge the subscriber list). All of the references to this PR are now confused as well. What benefit does a "clean" PR provide?

From my experience, a cleaner PR will be easier for new people to get the context or we look back after times. The more the comments are, the more confused readers get. Generally a cleaner PR is not suggested since it lost the context. But for this specific example, I think it makes sense since the context here is not so meaning full.

> Note that this leaves behind all who have hit "Subscribe" on this PR to keep tabs on things 

Sorry for that. But I think it may not be a big deal since people who still want to take an eye this can made it by clicking two times simply.

**Comment by mathstuf** [comment]
> Sorry for that. But I think it may not be a big deal since people who still want to take an eye this can made it by clicking two times simply.

The timing is also unfortunate as some may have already gone on vacation and will miss updates until they return. Anyways, what's done is done.

> The more the comments are, the more confused readers get. Generally a cleaner PR is not suggested since it lost the context.

It is unfortunate that Github sucks so much at rendering discussions (e.g., GItLab's threads that can be resolved makes it easy to collapse down finished things), There is the "mark as off-topic" strategy that I've seen used before, but that is also conflating the desire to hide things with spam control. In the future I would suggest mentioning those who have participated in the prior PR directly so that there's no gap in email/notification coverage.

**Comment by ChuanqiXu9** [comment]
Yeah, it'll be much better if we had that feature.

**Pull request closed without merging**